### ISSUE TITLE
Upgrade from realm 3.6.4 to 6.1.5 or 10.2.0 breaks schema configuration with es6 classes

### ISSUE THREAD

**Issue by angelos3lex** [realm/realm-js#3587]
## Goals
We are trying to upgrade to either latest realm or at least 6.1.5 so that we can use node 12 which is not supported in realm 3.6.4, in a react-native app.

## Expected Results
 - What we are trying is super minimal: Add two entries of `TableA` to the db, and then `getAll()` to print some info about each in the console.
 - We need to use es6 class schema declaration so that we can directly use business methods from our DbObjects.
 - The setup that was working in realm 3.6.4 to be working in latest v6 or v10 (or at least with some alterations)
 
## Actual Results
 With the exact same setup, only the first object of `TableA` is able to be saved. The second one can't be saved, because all its values can't be changed from null/undefined. We set its code to a valid uuidv4 string before saving, but it is still null on the next line, producing the error: `Error: TableA.code must be of type 'string', got 'null' (null)`
 
## Steps to Reproduce
I've made a public minimal repo [here](https://github.com/angelos3lex/RealmUpgradeTest), so that you can easily reproduce.
Just `npm install` and `run-android`. Press on the main button, and you will immediately see the crash on the logs with detailed info. The important files to look at are: [App.js](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/App.js), [MyRealm.js](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/MyRealm.js), [TableA.js](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/TableA.js), [TestDbModule.js](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/TestDbModule.js).

Just to cover some potential suggestions that we already tried:
- Changing our classes to extend Realm.Object has not at all any difference.
- Declaring the schema as `[TableA.schema]` instead of `[TableA]` is working, but the business methods of `TableA` models are not recognized. (i.e the `getInfo()` method of `TableA`)

## Version of Realm and Tooling

- Realm JS SDK Version: 6.1.5/10.2.0
- Node or React Native: React-Native (0.63.4), using either node 10.17.0 or 12.14.1
- Client OS & Version: Tested on Android (7.1.1 & 10) 
- Which debugger for React Native: None

**Comment by steffenagger** [comment]
Hello @angelos3lex

Yes, something is not working quite as intended with class models, when newing them for creation/insertion (https://github.com/realm/realm-js/issues/3553).

A workaround could be to use plain object literals, just for creation/insertion, realm will still return instances of the class models (I'm only suggesting this, to potentially unblock you guys).

**Comment by angelos3lex** [comment]
@steffenagger Thanks for your answer. Do you guys have this in your immediate to-dos or this may take months to be fixed? Thinking whether we should go on with the workaround or just delay a bit the update

**Comment by steffenagger** [comment]
> @steffenagger Thanks for your answer. Do you guys have this in your immediate to-dos or this may take months to be fixed? Thinking whether we should go on with the workaround or just delay a bit the update

@angelos3lex I'll know a bit more tomorrow, and will post an update.

**Comment by steffenagger** [comment]
I'll push for this to be prioritized, but that wont guarantee speed of delivery - our investigations could potentially trigger a discussions about assumptions and use cases, when it comes to class models.
Personally, I would opt in for the workaround (I'd rather that we surprise positively in the other direction).

When implementing the workaround, I would suggest implementing it by extending your existing models, to ensure object literals not floating around your codebase (I'm guessing this is also very much what you would like to avoid), and to allow for an refactor, like so:

```js
class Example extends Realm.Object {
  constructor(id, name, description) { /* left out */ }
  // existing business logic
  static schema = { /* left out */ }

  static insertionModel = (id, name, description) => ({
    id, name, description
  })
}
```

Later it should be a fairly straight forward refactor:
```js
// using the workaround
realm.write(() => {
  realm.create(Example, Example.insertionModel('123', 'example', 'this is an example'))
})

// back to using class model constructors again
realm.write(() => {
  realm.create(Example, new Example('123', 'example', 'this is an example'))
})
```

**Comment by angelos3lex** [comment]
hey @steffenagger 
I tried the workaround in the example repo that i had attached in the initial post. I still get the same error.
I've updated the repo with the changes mentioned in the workaround: [App.js](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/App.js), [TableA](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/TableA.js), [MyRealm.js](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/MyRealm.js)

**Comment by steffenagger** [comment]
> hey @steffenagger
> I tried the workaround in the example repo that i had attached in the initial post. I still get the same error.

@angelos3lex the issue is already when newing a class model, so it's already when doing `new TableA()`, the values in [MyRealm.js#L36](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/MyRealm.js#L36) will still be undefined.

For the workaround to work, you'll need to use `TableA.insertionModel(...)` instead of `new TableA(...)` in [App.js#L18](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/App.js#L18) and [App.js#L24](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/App.js#L24) (this will break the usage of `getInfo()` in  [MyRealm.js#L25-L30](https://github.com/angelos3lex/RealmUpgradeTest/blob/master/MyRealm.js#L25-L30))

**Comment by angelos3lex** [comment]
So the workaround is not an actual workaround for us, as the business methods are needed before the actual realm save as well.
We decided to wrap our db models to business objects and see how this will go.. Atm we experience some `Fatal signal 11` errors but we need to further investigate.

Is there any chance that realm 3.x versions are gonna be updated with the fix about 14.5 ios? Or only latest 6 and 10 versions will support it? Unfortunately our db is fairly big (more than 60 tables) and the effort for the upgrade is a big pain :S

**Comment by steffenagger** [comment]
> So the workaround is not an actual workaround for us, as the business methods are needed before the actual realm save as well.

I completely understand, we're currently investigating the forward approach of class models. I'll be sure to include this use case in our discussions.

> Is there any chance that realm 3.x versions are gonna be updated with the fix about 14.5 ios? Or only latest 6 and 10 versions will support it? Unfortunately our db is fairly big (more than 60 tables) and the effort for the upgrade is a big pain :S

Yeah only v6 and v10 will get updates going forward. But, maybe there is a bit of good news for you: Apple rolled back the change breaking Realm (or made exceptions for RN), in its latest beta (see: https://github.com/realm/realm-js/issues/3557#issuecomment-780799525).
So it might be worth checking your current solution against the latest beta first?

**Comment by hussainimp** [comment]
@steffenagger any update how long will this take?

**Comment by steffenagger** [comment]
> @steffenagger any update how long will this take?

@hussainimp It's high on our todo-list, but no, I don't have a timeframe for it.

**Comment by pramahaditamaputra-sicepat** [comment]
is there any update for this issue ?

**Comment by kneth** [comment]
@pramahaditamaputra-sicepat We are currently investigating/designing how our future support class-based models should look like. I can't give an exact time frame but it is an active project.

**Comment by kneth** [comment]
We have implemented - but not yet released - better support for classes. It is a breaking change and it will be released in version 11.0.0. We cannot give you an exact release date but we advice you to try out our prereleases of v11 as they come out.

See https://github.com/realm/realm-js/blob/v11/CHANGELOG.md for details.
